test/upgrade/fleet_test.go

```diff
@@ -0,0 +1,371 @@
+package upgrade
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/fleetdm/fleet/v4/server/service"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+)
+
+// Slots correspond to docker-compose fleet services, either fleet-a or fleet-b
+const (
+	slotA = "a"
+	slotB = "b"
+)
+
+func init() {
+	rand.Seed(time.Now().Unix())
+}
+
+// Fleet represents the fleet server and its dependencies used for testing.
+type Fleet struct {
+	ProjectName string
+	FilePath    string
+	Version     string
+	Token       string
+
+	dockerClient client.ContainerAPIClient
+}
+
+// NewFleet starts fleet and it's dependencies with the specified version.
+func NewFleet(t *testing.T, version string) *Fleet {
+	// don't use test name because it will be normalized
+	//nolint:gosec // does not need to be secure for tests
+	projectName := "fleet-test-" + strconv.FormatUint(rand.Uint64(), 16)
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		t.Fatalf("create docker client: %v", err)
+	}
+
+	f := &Fleet{
+		ProjectName:  projectName,
+		FilePath:     "docker-compose.yaml",
+		Version:      version,
+		dockerClient: dockerClient,
+	}
+
+	t.Cleanup(f.cleanup)
+
+	if err := f.Start(); err != nil {
+		t.Fatalf("start fleet: %v", err)
+	}
+
+	return f
+}
+
+func (f *Fleet) Start() error {
+	env := map[string]string{
+		"FLEET_VERSION_A": f.Version,
+	}
+	_, err := f.execCompose(env, "pull", "--parallel")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "fail to pull: %v\n", err)
+		fmt.Fprintf(os.Stderr, "retrying pull...\n")
+
+		_, err := f.execCompose(env, "pull", "--parallel")
```

Comment:
Curious why retrying only once.

Reply:
There wasn't really a good reason. Perhaps I should remove the retry entirely.